File: inspire/modules/workflows/actions/core_approval.py
# -*- coding: utf-8 -*-
#
# This file is part of INSPIRE.
# Copyright (C) 2014, 2015 CERN.
#
# INSPIRE is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# INSPIRE is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
#
# In applying this licence, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization
# or submit itself to any jurisdiction.

"""Approval action for INSPIRE."""

from flask import current_app, render_template

from invenio_accounts.models import User

from invenio_deposit.models import Deposition

from .hep_approval import hep_approval


class core_approval(hep_approval):

    """Class representing the approval action."""

    def render_mini(self, obj):
        """Method to render the minified action."""
        try:
            user = User.query.get(obj.id_user)
            d = Deposition(obj)
            rejection_text = "\n".join([line.strip() for line in render_template(
                'deposit/tickets/user_rejected.html',
                user=user,
                title=d.title
            ).split("\n")])
        except Exception:
            current_app.logger.exception("Failed to load rejection_text")
            rejection_text = ""
        return render_template(
            'workflows/actions/core_approval_mini.html',
            message=obj.get_action_message(),
            object=obj,
            resolve_url=self.url,
            rejection_text=rejection_text
        )

    def render(self, obj):
        """Method to render the action."""
        try:
            user = User.query.get(obj.id_user)
            d = Deposition(obj)
            rejection_text = "\n".join([line.strip() for line in render_template(
                'deposit/tickets/user_rejected.html',
                user=user,
                title=d.title
            ).split("\n")])
        except Exception:
            current_app.logger.exception("Failed to load rejection_text")
            rejection_text = ""
        return {
            "side": render_template('workflows/actions/core_approval_side.html',
                                    message=obj.get_action_message(),
                                    object=obj,
                                    resolve_url=self.url,
                                    rejection_text=rejection_text),
            "main": render_template('workflows/actions/core_approval_main.html',
                                    message=obj.get_action_message(),
                                    object=obj,
                                    resolve_url=self.url,
                                    rejection_text=rejection_text)
        }

File: inspirehep/modules/workflows/static/js/workflows/actions/core_approval.js
/*
 * This file is part of INSPIRE.
 * Copyright (C) 2013, 2014, 2015, 2016 CERN.
 *
 * INSPIRE is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * INSPIRE is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with INSPIRE; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
 */


define(
  [
    'jquery',
    'flight/lib/component',
  ],
  function(
    $,
    defineComponent) {

    "use strict";

    return defineComponent(CoreApprovalAction);

    /**
     * .. js:class:: CoreApprovalAction()
     *
     * Handles the events from the UI button elements for acceping/rejecting
     * records by sending the selected action to the server.
     *
     * The actionGroupSelector is required for handling display of any action
     * elements to be hidden or shown. It should be wrapped around every action
     * UI component.
     *
     * :param string actionResolveSelector: DOM selector of elements resolving actions.
     * :param string action_url: URL for resolving the action.
     *
     */
    function CoreApprovalAction() {

      this.attributes({
        actionAcceptSelector: ".core-approval-action-accept",
        actionRejectSelector: ".core-approval-action-reject",
        action_url: "",
        pdfCheckboxSelector: "[name='submission-data-pdf']"
      });

      // POST object init
      this.get_action_values = function(elem) {
        var actionValues = {
          "value": elem.data("value"),
          "objectids": []
        };

        actionValues.objectids.push(elem.data("objectid"));
        return actionValues;
      };

      this.get_pdf_submission_value = function() {
        return $(this.attr.pdfCheckboxSelector).prop("checked");
      };

      this.post_request = function(data, element) {
        console.log(data.message);
        this.trigger(document, "updateAlertMessage", {
          category: data.category,
          message: data.message
        });

        // Check if we are in the details page or
        // the main list, to trigger the right event
        if (document.URL.indexOf('/details/')) {
          this.trigger(document, "changePage");
        } else {
          this.trigger(document, "reloadHoldingPenTable");
        }
      };

      this.onAccept = function(ev, data) {
        var element = $(data.el);
        var payload = this.get_action_values(element);
        var pdf_submission = this.get_pdf_submission_value();
        if (pdf_submission) {
          payload.pdf_submission = pdf_submission;
        }
        this.requestToServer(payload, element);
      };

      // Request to sever
      this.requestToServer = function(payload, element) {
        var $this = this;

        jQuery.ajax({
          type: "POST",
          url: $this.attr.action_url,
          data: payload,
          success: function(data) {
            $this.post_request(data, element);
          }
        });
        this.pdf_submission_readonly();
      };

      this.preRejection = function(ev, data) {
        var element = $(data.el);
        var payload = this.get_action_values(element);
        console.log("GOT HERE");
        this.trigger(document, "loadRejectionModal", payload);
      };

      this.doRejection = function(ev, data) {
        data.value = "reject";
        var element = $("button.core-approval-action-reject[data-objectid=" + data.objectid + "]");
        this.requestToServer(data, element);
      };

      this.pdf_submission_readonly = function() {
        $(this.attr.pdfCheckboxSelector).prop("disabled", true);
      };

      this.after('initialize', function() {
        // Custom handlers
        this.on("click", {
          actionAcceptSelector: this.onAccept,
          actionRejectSelector: this.preRejection
        });
        this.on("rejectConfirmed", this.doRejection);

        if ($(this.attr.pdfCheckboxSelector).prop('disabled')) {
          this.pdf_submission_readonly();
        }
        console.log("Core approval init");
      });
    }
  }
);
